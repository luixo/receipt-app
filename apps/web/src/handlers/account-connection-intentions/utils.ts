import { TRPCError } from "@trpc/server";
import type { z } from "zod";

import { ACCOUNT_CONNECTIONS_INTENTIONS } from "~db/consts";
import type { AccountsId, UsersId } from "~db/models";
import type { Database } from "~db/types";
import type { emailSchema } from "~web/handlers/validation";

export const addConnectionIntention = async (
	database: Database,
	fromAccountId: AccountsId,
	user: { name: string },
	toEmail: z.infer<typeof emailSchema>,
	asUserId: UsersId,
) => {
	const targetAccount = await database
		.selectFrom("accounts")
		.select(["id", "avatarUrl"])
		.where("email", "=", toEmail.lowercase)
		.executeTakeFirst();
	if (!targetAccount) {
		throw new TRPCError({
			code: "NOT_FOUND",
			message: `Account with email "${toEmail.original}" does not exist.`,
		});
	}
	const connectedUser = await database
		.selectFrom("accounts")
		.where("accounts.email", "=", toEmail.lowercase)
		.innerJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", fromAccountId),
		)
		.select("users.name")
		.executeTakeFirst();
	if (connectedUser) {
		throw new TRPCError({
			code: "CONFLICT",
			message: `Account with email "${toEmail.original}" is already connected to user "${connectedUser.name}".`,
		});
	}
	const viceVersaIntention = await database
		.selectFrom("accountConnectionsIntentions")
		.select("userId")
		.where((eb) =>
			eb.and({
				accountId: targetAccount.id,
				targetAccountId: fromAccountId,
			}),
		)
		.executeTakeFirst();
	if (viceVersaIntention) {
		await database.transaction().execute(async (tx) => {
			await tx
				.updateTable("users")
				.set({ connectedAccountId: fromAccountId })
				.where((eb) =>
					eb.and({
						ownerAccountId: targetAccount.id,
						id: viceVersaIntention.userId,
					}),
				)
				.executeTakeFirst();
			await tx
				.updateTable("users")
				.set({ connectedAccountId: targetAccount.id })
				.where((eb) =>
					eb.and({
						ownerAccountId: fromAccountId,
						id: asUserId,
					}),
				)
				.executeTakeFirst();
			await tx
				.deleteFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: targetAccount.id,
						targetAccountId: fromAccountId,
					}),
				)
				.executeTakeFirst();
		});
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: true,
			user: {
				name: user.name,
			},
		};
	}
	try {
		await database
			.insertInto("accountConnectionsIntentions")
			.values({
				accountId: fromAccountId,
				targetAccountId: targetAccount.id,
				userId: asUserId,
				createdAt: new Date(),
			})
			.executeTakeFirst();
		return {
			account: {
				id: targetAccount.id,
				email: toEmail.lowercase,
				avatarUrl: targetAccount.avatarUrl || undefined,
			},
			connected: false,
			user: {
				name: user.name,
			},
		};
	} catch (e) {
		// Could be like `duplicate key value violates unique constraint "..."`
		const message = String(e);
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.ACCOUNT_PAIR)
		) {
			const existingIntention = await database
				.selectFrom("accountConnectionsIntentions")
				.where((eb) =>
					eb.and({
						accountId: fromAccountId,
						targetAccountId: targetAccount.id,
					}),
				)
				.select(["userId"])
				.executeTakeFirstOrThrow();
			const existingUser = await database
				.selectFrom("users")
				.select("name")
				.where("id", "=", existingIntention.userId)
				.executeTakeFirstOrThrow();
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${toEmail.original}" as user "${existingUser.name}".`,
			});
		}
		if (
			message.includes(ACCOUNT_CONNECTIONS_INTENTIONS.CONSTRAINTS.USER_PAIR)
		) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${user.name}".`,
			});
			// This is probably a c8 bug
			/* c8 ignore next */
		}
		/* c8 ignore next 2 */
		throw e;
	}
};
