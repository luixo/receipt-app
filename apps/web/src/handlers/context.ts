import { type H3Event, createEvent } from "@tanstack/react-start/server";
import type { inferProcedureBuilderResolverOptions } from "@trpc/server";
import type { TRPCRequestInfo } from "@trpc/server/http";
import type { IncomingMessage, ServerResponse } from "node:http";

import type { Database } from "~db/database";
import type { TestContext } from "~tests/backend/utils/test";
import type { authProcedure } from "~web/handlers/trpc";
import type { CacheDbOptions } from "~web/providers/cache-db";
import type { EmailOptions } from "~web/providers/email";
import type { ExchangeRateOptions } from "~web/providers/exchange-rate";
import type { Logger } from "~web/providers/logger";
import type { S3Options } from "~web/providers/s3";

type TestContextPicks = Pick<TestContext, "getSalt" | "getUuid"> & {
	database: Database;
	logger: Logger;
	emailOptions: EmailOptions;
	cacheDbOptions: CacheDbOptions;
	exchangeRateOptions: ExchangeRateOptions;
	s3Options: S3Options;
};

export type NetContext = {
	event: H3Event;
	info: TRPCRequestInfo;
};

export type UnauthorizedContext = NetContext & TestContextPicks;

export type AuthorizedContext = inferProcedureBuilderResolverOptions<
	typeof authProcedure
>["ctx"];

export const createContext = (
	{
		req,
		res,
		info,
	}: { req: IncomingMessage; res: ServerResponse; info: TRPCRequestInfo },
	testContext: TestContextPicks,
): UnauthorizedContext => ({
	info,
	event: createEvent(req, res),
	...testContext,
});
