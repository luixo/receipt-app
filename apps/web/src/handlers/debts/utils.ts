import type { Database, DebtsId, SimpleInsertObject } from "~db";
import { nonNullishGuard } from "~utils";

export const upsertAutoAcceptedDebts = async (
	database: Database,
	debts: (SimpleInsertObject<"debts"> & { isNew: boolean })[],
) => {
	const fetchedDebts = await database
		.selectFrom("debts")
		.where((eb) =>
			eb.or(
				debts.map((debt) =>
					!debt.isNew
						? eb.and({
								id: debt.id,
								ownerAccountId: debt.ownerAccountId,
						  })
						: debt.receiptId
						? eb.and({
								ownerAccountId: debt.ownerAccountId,
								userId: debt.userId,
								receiptId: debt.receiptId,
						  })
						: eb("debts.ownerAccountId", "is", null),
				),
			),
		)

		.select([
			"debts.id",
			"debts.ownerAccountId",
			"debts.userId",
			"debts.receiptId",
		])
		.execute();
	const counterpartyDebts = debts.map((debt) =>
		fetchedDebts.find((fetchedDebt) => {
			if (debt.isNew) {
				if (!debt.receiptId || !fetchedDebt.receiptId) {
					return false;
				}
				return (
					fetchedDebt.ownerAccountId === debt.ownerAccountId &&
					fetchedDebt.userId === debt.userId &&
					fetchedDebt.receiptId === debt.receiptId
				);
			}
			return (
				fetchedDebt.ownerAccountId === debt.ownerAccountId &&
				fetchedDebt.id === debt.id
			);
		}),
	);
	const existentDebts = debts
		.map((nextDebt, index) =>
			counterpartyDebts[index]
				? ([nextDebt, counterpartyDebts[index]!] as const)
				: null,
		)
		.filter(nonNullishGuard);
	const nonExistentDebts = debts
		.map((nextDebt, index) => (counterpartyDebts[index] ? null : nextDebt))
		.filter(nonNullishGuard);
	const [newDebts, ...updatedDebts] = await Promise.all([
		nonExistentDebts.length === 0
			? ([] as { id: DebtsId }[])
			: database
					.insertInto("debts")
					.values(nonExistentDebts.map(({ isNew, ...debt }) => debt))
					.returning(["debts.id"])
					.execute(),
		...existentDebts.map(([nextDebt, currentDebt]) =>
			database
				.updateTable("debts")
				.set({
					amount: nextDebt.amount,
					timestamp: nextDebt.timestamp,
					currencyCode: nextDebt.currencyCode,
					receiptId: nextDebt.receiptId,
					lockedTimestamp: nextDebt.lockedTimestamp,
				})
				.where((eb) =>
					eb.and({
						id: currentDebt.id,
						ownerAccountId: currentDebt.ownerAccountId,
						userId: currentDebt.userId,
					}),
				)
				.returning(["debts.id", "debts.receiptId"])
				.executeTakeFirst(),
		),
	]);
	return { newDebts, updatedDebts };
};
