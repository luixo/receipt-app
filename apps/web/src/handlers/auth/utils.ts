import { TRPCError } from "@trpc/server";

import { DAY } from "~utils";
import type { AccountsId, SessionsSessionId } from "~web/db/models";
import { generateConfirmEmailEmail } from "~web/email/utils";
import type { UnauthorizedContext } from "~web/handlers/context";
import { getEmailClient } from "~web/providers/email";

// How long a session should last
export const SESSION_EXPIRATION_DURATION = 30 * DAY;
// How long session should be intact before we update it
export const SESSION_SHOULD_UPDATE_EVERY = 2 * DAY;

export const getExpirationDate = () =>
	new Date(Date.now() + SESSION_EXPIRATION_DURATION);

export const createAuthorizationSession = async (
	ctx: UnauthorizedContext,
	accountId: AccountsId,
) => {
	const uuid: SessionsSessionId = ctx.getUuid();
	const expirationDate = getExpirationDate();
	await ctx.database
		.insertInto("sessions")
		.values({
			accountId,
			sessionId: uuid,
			expirationTimestamp: expirationDate,
		})
		.executeTakeFirst();
	return {
		authToken: uuid,
		expirationDate,
	};
};

export const sendVerificationEmail = async (
	ctx: UnauthorizedContext,
	email: string,
	token: string,
) => {
	if (!ctx.emailOptions.active) {
		throw new TRPCError({
			code: "FORBIDDEN",
			message: "Currently email resend is not supported",
		});
	}
	try {
		await getEmailClient(ctx).send({
			address: email,
			subject: "Confirm email in Receipt App",
			body: generateConfirmEmailEmail(ctx, token),
		});
	} catch (e) {
		throw new TRPCError({
			code: "INTERNAL_SERVER_ERROR",
			message: `Something went wrong: ${
				/* c8 ignore next */
				e instanceof Error ? e.message : String(e)
			}`,
		});
	}
};
