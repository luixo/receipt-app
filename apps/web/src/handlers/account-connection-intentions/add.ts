import { TRPCError } from "@trpc/server";
import { entries } from "remeda";
import { z } from "zod";

import type { AccountsId } from "~db/models";
import { queueCallFactory } from "~web/handlers/batch";
import type { AuthorizedContext } from "~web/handlers/context";
import { authProcedure } from "~web/handlers/trpc";
import { emailSchema, userIdSchema } from "~web/handlers/validation";

const addConnectionIntentionSchema = z.strictObject({
	userId: userIdSchema,
	email: emailSchema,
});
type ConnectionIntention = z.infer<typeof addConnectionIntentionSchema>;

const getTargetUsers = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) =>
	ctx.database
		.selectFrom("users")
		.where(
			"users.id",
			"in",
			intentions.map((intention) => intention.userId),
		)
		.leftJoin("accounts", (qb) =>
			qb.onRef("accounts.id", "=", "users.connectedAccountId"),
		)
		.select([
			"users.id",
			"users.name",
			"accounts.email",
			"users.ownerAccountId",
		])
		.execute();

const getTargetAccounts = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) =>
	ctx.database
		.selectFrom("accounts")
		.where(
			"accounts.email",
			"in",
			intentions.map((intention) => intention.email.lowercase),
		)
		.leftJoin("users", (qb) =>
			qb
				.onRef("users.connectedAccountId", "=", "accounts.id")
				.on("users.ownerAccountId", "=", ctx.auth.accountId),
		)
		.select([
			"accounts.id",
			"accounts.avatarUrl",
			"accounts.email",
			"users.name as userName",
		])
		.execute();

const getDirectIntentions = async (
	ctx: AuthorizedContext,
	targetAccountsPromise: ReturnType<typeof getTargetAccounts>,
	targetUsersPromise: ReturnType<typeof getTargetUsers>,
) => {
	const targetAccounts = await targetAccountsPromise;
	const targetUsers = await targetUsersPromise;
	if (targetAccounts.length === 0 || targetUsers.length === 0) {
		return [];
	}
	return ctx.database
		.selectFrom("accountConnectionsIntentions")
		.where("accountConnectionsIntentions.accountId", "=", ctx.auth.accountId)
		.where((qb) =>
			qb.or([
				qb(
					"accountConnectionsIntentions.targetAccountId",
					"in",
					targetAccounts.map(({ id }) => id),
				),
				qb(
					"accountConnectionsIntentions.userId",
					"in",
					targetUsers.map(({ id }) => id),
				),
			]),
		)
		.innerJoin("users", (qb) =>
			qb.onRef("users.id", "=", "accountConnectionsIntentions.userId"),
		)
		.select([
			"accountConnectionsIntentions.userId",
			"accountConnectionsIntentions.targetAccountId",
			"users.name",
		])
		.execute();
};

const getViceVersaIntentions = async (
	ctx: AuthorizedContext,
	targetAccountsPromise: ReturnType<typeof getTargetAccounts>,
) => {
	const targetAccounts = await targetAccountsPromise;
	if (targetAccounts.length === 0) {
		return [];
	}
	return ctx.database
		.selectFrom("accountConnectionsIntentions")
		.where(
			"accountConnectionsIntentions.accountId",
			"in",
			targetAccounts.map(({ id }) => id),
		)
		.where(
			"accountConnectionsIntentions.targetAccountId",
			"=",
			ctx.auth.accountId,
		)
		.select([
			"accountConnectionsIntentions.userId",
			"accountConnectionsIntentions.accountId",
		])
		.execute();
};

const getData = async (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
) => {
	const targetAccountsPromise = getTargetAccounts(ctx, intentions);
	const targetUsersPromise = getTargetUsers(ctx, intentions);

	return {
		targetAccounts: await targetAccountsPromise,
		targetUsers: await targetUsersPromise,
		directIntentions: await getDirectIntentions(
			ctx,
			targetAccountsPromise,
			targetUsersPromise,
		),
		viceVersaIntentions: await getViceVersaIntentions(
			ctx,
			targetAccountsPromise,
		),
	};
};

const getIntentionsOrErrors = (
	ctx: AuthorizedContext,
	intentions: readonly ConnectionIntention[],
	{
		targetAccounts,
		targetUsers,
		viceVersaIntentions,
		directIntentions,
	}: Awaited<ReturnType<typeof getData>>,
) =>
	intentions.map((intention) => {
		const targetUser = targetUsers.find((user) => user.id === intention.userId);
		if (!targetUser) {
			return new TRPCError({
				code: "NOT_FOUND",
				message: `User "${intention.userId}" does not exist.`,
			});
		}
		if (targetUser.ownerAccountId !== ctx.auth.accountId) {
			return new TRPCError({
				code: "FORBIDDEN",
				message: `User "${intention.userId}" is not owned by "${ctx.auth.email}".`,
			});
		}
		if (targetUser.email) {
			return new TRPCError({
				code: "CONFLICT",
				message: `User "${intention.userId}" is already connected to account "${targetUser.email}".`,
			});
		}
		const targetAccount = targetAccounts.find(
			(account) => account.email === intention.email.lowercase,
		);
		if (!targetAccount) {
			return new TRPCError({
				code: "NOT_FOUND",
				message: `Account with email "${intention.email.original}" does not exist.`,
			});
		}
		if (targetAccount.userName) {
			return new TRPCError({
				code: "CONFLICT",
				message: `Account with email "${intention.email.original}" is already connected to user "${targetAccount.userName}".`,
			});
		}
		const directIntentionByAccount = directIntentions.find(
			({ targetAccountId }) => targetAccountId === targetAccount.id,
		);
		if (directIntentionByAccount) {
			return new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to "${intention.email.original}" as user "${directIntentionByAccount.name}".`,
			});
		}
		const directIntentionByUser = directIntentions.find(
			({ userId }) => userId === intention.userId,
		);
		if (directIntentionByUser) {
			return new TRPCError({
				code: "CONFLICT",
				message: `You already has intention to connect to user "${directIntentionByUser.name}".`,
			});
		}
		const viceVersaIntention = viceVersaIntentions.find(
			({ accountId }) => accountId === targetAccount.id,
		);
		if (viceVersaIntention) {
			return {
				type: "vice-versa" as const,
				targetAccount: {
					id: targetAccount.id,
					email: targetAccount.email,
					avatarUrl: targetAccount.avatarUrl,
				},
				asUser: {
					id: intention.userId,
					name: targetUser.name,
				},
				viceVersaUser: {
					id: viceVersaIntention.userId,
				},
			};
		}
		return {
			type: "direct" as const,
			targetAccount: {
				id: targetAccount.id,
				email: targetAccount.email,
				avatarUrl: targetAccount.avatarUrl,
			},
			asUser: {
				id: intention.userId,
				name: targetUser.name,
			},
		};
	});

type Intention = Exclude<
	ReturnType<typeof getIntentionsOrErrors>[number],
	TRPCError
>;

const insertViceVersaIntentions = async (
	ctx: AuthorizedContext,
	intentions: Extract<Intention, { type: "vice-versa" }>[],
) => {
	if (intentions.length === 0) {
		return;
	}
	await ctx.database.transaction().execute((tx) =>
		Promise.all([
			...intentions.map((intention) =>
				tx
					.updateTable("users")
					.set({ connectedAccountId: ctx.auth.accountId })
					.where((eb) =>
						eb.and({
							ownerAccountId: intention.targetAccount.id,
							id: intention.viceVersaUser.id,
						}),
					)
					.executeTakeFirst(),
			),
			...intentions.map((intention) =>
				tx
					.updateTable("users")
					.set({ connectedAccountId: intention.targetAccount.id })
					.where((eb) =>
						eb.and({
							ownerAccountId: ctx.auth.accountId,
							id: intention.asUser.id,
						}),
					)
					.executeTakeFirst(),
			),
			tx
				.deleteFrom("accountConnectionsIntentions")
				.where(
					"accountConnectionsIntentions.targetAccountId",
					"=",
					ctx.auth.accountId,
				)
				.where(
					"accountConnectionsIntentions.accountId",
					"in",
					intentions.map((intention) => intention.targetAccount.id),
				)
				.executeTakeFirst(),
		]),
	);
};

const insertDirectIntentions = async (
	ctx: AuthorizedContext,
	intentions: Extract<Intention, { type: "direct" }>[],
) => {
	if (intentions.length === 0) {
		return;
	}
	await ctx.database
		.insertInto("accountConnectionsIntentions")
		.values(
			intentions.map((intention) => ({
				accountId: ctx.auth.accountId,
				targetAccountId: intention.targetAccount.id,
				userId: intention.asUser.id,
				createdAt: new Date(),
			})),
		)
		.execute();
};

const getDuplicates = <T, K extends string>(
	array: readonly T[],
	getKey: (item: T) => K,
) =>
	entries(
		array
			.map(getKey)
			.reduce<Partial<Record<string, number>>>(
				(acc, element) => ({ ...acc, [element]: (acc[element] ?? 0) + 1 }),
				{},
			),
	).filter(([, value]) => value !== 1);

type IntentionOutput = {
	account: {
		id: AccountsId;
		email: string;
		avatarUrl?: string;
	};
	connected: boolean;
	user: {
		name: string;
	};
};

export const batchFn =
	(ctx: AuthorizedContext) =>
	async (
		inputs: readonly z.infer<typeof addConnectionIntentionSchema>[],
	): Promise<(IntentionOutput | TRPCError)[]> => {
		const duplicatedEmails = getDuplicates(
			inputs,
			(intention) => intention.email.lowercase,
		);
		if (duplicatedEmails.length !== 0) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `Expected to have unique emails, got repeating emails: ${duplicatedEmails
					.map(([email, count]) => `"${email}" (${count} times)`)
					.join(", ")}.`,
			});
		}
		const duplicatedUserIds = getDuplicates(
			inputs,
			(intention) => intention.userId,
		);
		if (duplicatedUserIds.length !== 0) {
			throw new TRPCError({
				code: "CONFLICT",
				message: `Expected to have unique user ids, got repeating: ${duplicatedUserIds
					.map(([userId, count]) => `"${userId}" (${count} times)`)
					.join(", ")}.`,
			});
		}
		const data = await getData(ctx, inputs);
		const intentionsOrErrors = getIntentionsOrErrors(ctx, inputs, data);
		const intentions = intentionsOrErrors.filter(
			(
				intentionOrError,
			): intentionOrError is Exclude<typeof intentionOrError, TRPCError> =>
				!(intentionOrError instanceof TRPCError),
		);
		const directIntentions = intentions.filter(
			(intention): intention is Extract<typeof intention, { type: "direct" }> =>
				intention.type === "direct",
		);
		const viceVersaIntentions = intentions.filter(
			(
				intention,
			): intention is Extract<typeof intention, { type: "vice-versa" }> =>
				intention.type === "vice-versa",
		);
		await Promise.all([
			insertDirectIntentions(ctx, directIntentions),
			insertViceVersaIntentions(ctx, viceVersaIntentions),
		]);
		return intentionsOrErrors.map((intentionOrError) => {
			if (intentionOrError instanceof TRPCError) {
				return intentionOrError;
			}
			const matchedDirectIntention = directIntentions.find(
				({ targetAccount }) =>
					targetAccount.id === intentionOrError.targetAccount.id,
			);
			const matchedViceVersaIntention = viceVersaIntentions.find(
				({ targetAccount }) =>
					targetAccount.id === intentionOrError.targetAccount.id,
			);
			/* c8 ignore start */
			if (!matchedViceVersaIntention && !matchedDirectIntention) {
				return new TRPCError({
					code: "INTERNAL_SERVER_ERROR",
					message: `Expected to have an intention in either direct or vice-versa connection list.`,
				});
			}
			/* c8 ignore stop */
			return {
				account: {
					id: intentionOrError.targetAccount.id,
					email: intentionOrError.targetAccount.email,
					avatarUrl: intentionOrError.targetAccount.avatarUrl || undefined,
				},
				connected: Boolean(matchedViceVersaIntention),
				user: {
					name: intentionOrError.asUser.name,
				},
			};
		});
	};

const queueAddConnectionIntenion = queueCallFactory<
	AuthorizedContext,
	z.infer<typeof addConnectionIntentionSchema>,
	IntentionOutput
>(batchFn);

export const procedure = authProcedure
	.input(addConnectionIntentionSchema)
	.mutation(queueAddConnectionIntenion);
