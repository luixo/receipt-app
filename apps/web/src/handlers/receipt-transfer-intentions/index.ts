import { t } from "~web/handlers/trpc";

import { procedure as accept } from "./accept";
import { procedure as add } from "./add";
import { procedure as getAll } from "./get-all";
import { procedure as remove } from "./remove";

export const router = t.router({
	getAll,
	add,
	remove,
	accept,
});
