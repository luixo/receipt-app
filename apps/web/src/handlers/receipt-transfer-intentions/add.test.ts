import { faker } from "@faker-js/faker";
import { describe, expect } from "vitest";

import { createAuthContext } from "~tests/backend/utils/context";
import {
	insertAccount,
	insertAccountWithSession,
	insertConnectedUsers,
	insertReceipt,
	insertReceiptParticipant,
	insertUser,
} from "~tests/backend/utils/data";
import {
	expectDatabaseDiffSnapshot,
	expectTRPCError,
	expectUnauthorizedError,
} from "~tests/backend/utils/expect";
import { test } from "~tests/backend/utils/test";
import { t } from "~web/handlers/trpc";

import { procedure } from "./add";

const createCaller = t.createCallerFactory(t.router({ procedure }));

describe("receiptTransferIntentions.add", () => {
	describe("input verification", () => {
		expectUnauthorizedError((context) =>
			createCaller(context).procedure({
				targetEmail: faker.internet.email(),
				receiptId: faker.string.uuid(),
			}),
		);

		describe("receiptId", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							targetEmail: faker.internet.email(),
							receiptId: "invalid-uuid",
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "receiptId": Invalid uuid`,
				);
			});
		});

		describe("targetEmail", () => {
			test("invalid", async ({ ctx }) => {
				const { sessionId } = await insertAccountWithSession(ctx);

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							targetEmail: "invalid-email",
							receiptId: faker.string.uuid(),
						}),
					"BAD_REQUEST",
					`Zod error\n\nAt "targetEmail": Invalid email address`,
				);
			});
		});

		test("targeting yourself", async ({ ctx }) => {
			const {
				sessionId,
				accountId,
				account: { email },
			} = await insertAccountWithSession(ctx);
			const { id: foreignAccountId } = await insertAccount(ctx);
			await insertConnectedUsers(ctx, [accountId, foreignAccountId]);
			const { id: receiptId } = await insertReceipt(ctx, accountId);

			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						targetEmail: email,
						receiptId,
					}),
				"BAD_REQUEST",
				`Cannot transfer receipt to yourself`,
			);
		});

		describe("target account", () => {
			test("does not exist", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: receiptId } = await insertReceipt(ctx, accountId);
				const fakeEmail = faker.internet.email();

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							targetEmail: fakeEmail,
							receiptId,
						}),
					"BAD_REQUEST",
					`Account "${fakeEmail}" does not exist or you don't have a user connected to it.`,
				);
			});

			test("is not connected with your account", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: receiptId } = await insertReceipt(ctx, accountId);
				const { email } = await insertAccount(ctx);

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							targetEmail: email,
							receiptId,
						}),
					"BAD_REQUEST",
					`Account "${email}" does not exist or you don't have a user connected to it.`,
				);
			});
		});

		test("receipt does not exist", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: foreignAccountId, email: foreignEmail } = await insertAccount(
				ctx,
			);
			await insertConnectedUsers(ctx, [accountId, foreignAccountId]);
			const fakeReceiptId = faker.string.uuid();

			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						targetEmail: foreignEmail,
						receiptId: fakeReceiptId,
					}),
				"NOT_FOUND",
				`Receipt "${fakeReceiptId}" does not exist or you are not its owner.`,
			);
		});

		test("receipt is not owned by an account", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: foreignAccountId, email: foreignEmail } = await insertAccount(
				ctx,
			);
			const [, { id: foreignToSelfUserId }] = await insertConnectedUsers(ctx, [
				accountId,
				foreignAccountId,
			]);
			const { id: foreignReceiptId } = await insertReceipt(
				ctx,
				foreignAccountId,
			);
			// Being added to the receipt doesn't allow you to create transfer intentions
			await insertReceiptParticipant(
				ctx,
				foreignReceiptId,
				foreignToSelfUserId,
			);
			// Own receipts don't interfere with this error
			await insertReceipt(ctx, accountId);

			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						targetEmail: foreignEmail,
						receiptId: foreignReceiptId,
					}),
				"NOT_FOUND",
				`Receipt "${foreignReceiptId}" does not exist or you are not its owner.`,
			);
		});

		describe("receipt has users", () => {
			test("with connected account", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: foreignAccountId, email: foreignEmail } =
					await insertAccount(ctx);
				const [{ id: foreignUserId, name: foreignUserName }] =
					await insertConnectedUsers(ctx, [accountId, foreignAccountId]);
				const { id: receiptId } = await insertReceipt(ctx, accountId);
				await insertReceiptParticipant(ctx, receiptId, foreignUserId);

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							targetEmail: foreignEmail,
							receiptId,
						}),
					"BAD_REQUEST",
					`Cannot send receipt transfer intention while having user "${foreignUserName}" added to it.`,
				);
			});

			test("without connected account", async ({ ctx }) => {
				const { sessionId, accountId } = await insertAccountWithSession(ctx);
				const { id: foreignAccountId, email: foreignEmail } =
					await insertAccount(ctx);
				await insertConnectedUsers(ctx, [accountId, foreignAccountId]);
				const { id: userId, name: userName } = await insertUser(ctx, accountId);
				const { id: receiptId } = await insertReceipt(ctx, accountId);
				await insertReceiptParticipant(ctx, receiptId, userId);

				const caller = createCaller(createAuthContext(ctx, sessionId));
				await expectTRPCError(
					() =>
						caller.procedure({
							targetEmail: foreignEmail,
							receiptId,
						}),
					"BAD_REQUEST",
					`Cannot send receipt transfer intention while having user "${userName}" added to it.`,
				);
			});
		});

		test("receipt is locked", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: foreignAccountId, email: foreignEmail } = await insertAccount(
				ctx,
			);
			await insertConnectedUsers(ctx, [accountId, foreignAccountId]);
			const { id: receiptId } = await insertReceipt(ctx, accountId, {
				lockedTimestamp: new Date(),
			});

			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						targetEmail: foreignEmail,
						receiptId,
					}),
				"BAD_REQUEST",
				`Cannot send receipt transfer intention while receipt is locked.`,
			);
		});

		test("receipt already has a transfer intention", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: foreignAccountId, email: foreignEmail } = await insertAccount(
				ctx,
			);
			const [{ name: foreignUserName }] = await insertConnectedUsers(ctx, [
				accountId,
				foreignAccountId,
			]);
			const { id: receiptId } = await insertReceipt(ctx, accountId, {
				transferIntentionAccountId: foreignAccountId,
			});

			const caller = createCaller(createAuthContext(ctx, sessionId));
			await expectTRPCError(
				() =>
					caller.procedure({
						targetEmail: foreignEmail,
						receiptId,
					}),
				"BAD_REQUEST",
				`There is already a transfer intention to user "${foreignUserName}".`,
			);
		});
	});

	describe("functionality", () => {
		test("transfer intention is added", async ({ ctx }) => {
			const { sessionId, accountId } = await insertAccountWithSession(ctx);
			const { id: foreignAccountId, email: foreignEmail } = await insertAccount(
				ctx,
			);
			await insertConnectedUsers(ctx, [accountId, foreignAccountId]);
			const { id: receiptId } = await insertReceipt(ctx, accountId);

			// Verify unrelated data doesn't affect the result
			const { id: anotherReceiptId } = await insertReceipt(ctx, accountId);
			const { id: anotherUserId } = await insertUser(ctx, accountId);
			await insertReceiptParticipant(ctx, anotherReceiptId, anotherUserId);

			const caller = createCaller(createAuthContext(ctx, sessionId));
			const result = await expectDatabaseDiffSnapshot(ctx, () =>
				caller.procedure({
					targetEmail: foreignEmail,
					receiptId,
				}),
			);
			expect(result).toBeUndefined();
		});
	});
});
