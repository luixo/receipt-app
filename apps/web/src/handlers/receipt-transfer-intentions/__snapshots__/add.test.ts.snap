// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`receiptTransferIntentions.add > functionality > transfer intention is added 1`] = `
{
  "updated": {
    "receipts": {
      "id:a5a9a69e-ad0f-404b-b6de-75bbc69db351": {
        "transferIntentionAccountId": "e9140e99-b5f6-426f-9023-3ba138cbb448",
        "updatedAt": "2020-01-01T00:01:00.000Z",
      },
    },
  },
}
`;
